File: src/Boundaries/reflectingALEMeshBoundary.cc
// Copyright (C) 2026  Cody Raskin

#pragma once
#include "aleMeshBoundary.hh"

// Solid-wall boundary: mirror the normal velocity component across the face,
// zero-gradient density/energy. Same physical convention as
// ReflectingGridBoundary's face-mode vector handling.
template <int dim>
class ReflectingALEMeshBoundary : public ALEMeshBoundary<dim> {
public:
    using Vector = Lin::Vector<dim>;

    ReflectingALEMeshBoundary(Mesh::ALEMesh<dim>* mesh) : ALEMeshBoundary<dim>(mesh) {}

    virtual void
    ghostState(size_t faceIndex,
               double rhoIn, const Vector& vIn, double uIn,
               double& rhoOut, Vector& vOut, double& uOut) const override {
        const auto& face = this->mesh->getFaces()[faceIndex];
        double vn = vIn.dot(face.normal);
        vOut = vIn - face.normal * (2.0 * vn);
        rhoOut = rhoIn;
        uOut = uIn;
    }
};

File: src/Boundaries/reflectingALEMeshBoundary.py
# Copyright (C) 2026  Cody Raskin

from PYB11Generator import *
from boundary import *

@PYB11template("dim")
class ReflectingALEMeshBoundary(Boundary):
    def pyinit(self, mesh="Mesh::ALEMesh<%(dim)s>*"):
        return
    def setFaces(self, ids="std::vector<size_t>"):
        return
    def setAllBoundaryFaces(self):
        return

ReflectingALEMeshBoundary2d = PYB11TemplateClass(ReflectingALEMeshBoundary,
                              template_parameters = ("2"),
                              cppname = "ReflectingALEMeshBoundary<2>",
                              pyname = "ReflectingALEMeshBoundary2d",
                              docext = " (2D).")
